src/bokeh/core/property/bases.py

```diff
@@ -338,6 +338,12 @@ def _hinted_value(self, value: Any, hint: DocumentPatchedEvent | None) -> Any:
         return value
 
     def prepare_value(self, owner: HasProps | type[HasProps], name: str, value: Any, *, hint: DocumentPatchedEvent | None = None) -> T:
+        import narwhals.stable.v1 as nw
+
+        # Try converting to Narwhals so that any dataframe supported by Narwhals
+        # can be supported.
+        value = nw.from_native(value, eager_only=True, allow_series=True, pass_through=True)
```

Comment:
@MarcoGorelli apologies I could have been more specific. My Idea was to create property types for lazy data frames and series, e.g. `LazyDataFrame` and `LazySeries` so that all the narwal specific code could neatly go in those (i.e. in their `prepare_value` and `validate` methods, etc.) As before with `Object`m this location still seems too low-level and general to involve `narwhals` checks, this `nw.from_native` call would now be invoked on every property use anywhere. If you already tried making `LazyDataFrame` and `LazySeries`, did you run into issues we can help with?

Reply:
Or, perhaps the way you have things now, this `prepare_value` can just move to `EagerDataFrame` ? I think your sense of `EagerDataFrame` (i.e. "*can be converted to* an eager data frame") is maybe what I meant by `LazyDataFrame` (I just approached from a different linguistic angle...) If it works to move this `prepare_value` there then I think that is all that is necessary.